### ISSUE TITLE
does not build with PGI

### ISSUE THREAD

**Issue by nikizadehgfdl** [oceans4/MOM_remapping_gpu_2020#16]
I am trying to test this on our gpu box and with PGI20 and it does not build.
I can easily fix the dumpGrid error ( use MOM_remapping, only : dumpGrid).
But I don't know what the R_X86_64_32S errors are about. They were not there in my original driver on this machine.

```
lscgpu50-d: /home/Niki.Zadeh/storage/MOM_remapping_gpu_2020/build % ../configure
checking whether the Fortran compiler works... yes
checking for Fortran compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU Fortran compiler... no
checking whether /opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran accepts -g... yes
checking for Fortran flag to compile .f90 files... none
checking for /opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran option to force 8-byte reals... -r8
configure: creating ./config.status
config.status: creating Makefile
lscgpu50-d: /home/Niki.Zadeh/storage/MOM_remapping_gpu_2020/build % make
/opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran -g -r8 -c ../src/MOM_error_handler.F90
/opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran -g -r8 -c ../src/regrid_solvers.F90
/opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran -g -r8 -c ../src/polynomial_functions.F90
/opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran -g -r8 -c ../src/regrid_edge_values.F90
/opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran -g -r8 -c ../src/PLM_functions.F90
/opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran -g -r8 -c ../src/MOM_string_functions.F90
/opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran -g -r8 -c ../src/MOM_remapping.F90
/opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran -g -r8 -c ../src/prg_unit_tests.F90
/opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran  -o prg_unit_tests prg_unit_tests.o MOM_remapping.o MOM_error_handler.o MOM_string_functions.o regrid_edge_values.o PLM_functions.o regrid_solvers.o polynomial_functions.o
/opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran -g -r8 -c ../src/prg_remap.F90
/opt/nvidia/hpc_sdk/Linux_x86_64/20.9/compilers/bin/nvfortran  -o prg_remap prg_remap.o MOM_remapping.o MOM_error_handler.o MOM_string_functions.o regrid_edge_values.o PLM_functions.o regrid_solvers.o polynomial_functions.o
prg_remap.o: In function `MAIN_':
/storage/nnz/MOM_remapping_gpu_2020/build/../src/prg_remap.F90:45: undefined reference to `dumpgrid_'
/storage/nnz/MOM_remapping_gpu_2020/build/../src/prg_remap.F90:53: undefined reference to `dumpgrid_'
MOM_remapping.o: In function `mom_remapping_remapping_unit_tests_':
/storage/nnz/MOM_remapping_gpu_2020/build/../src/MOM_remapping.F90:1618:(.text+0xb5e9): relocation truncated to fit: R_X86_64_PC32 against `.bss'
/storage/nnz/MOM_remapping_gpu_2020/build/../src/MOM_remapping.F90:1618:(.text+0xb5f7): relocation truncated to fit: R_X86_64_PC32 against `.bss'
/storage/nnz/MOM_remapping_gpu_2020/build/../src/MOM_remapping.F90:1618:(.text+0xb605): relocation truncated to fit: R_X86_64_PC32 against `.bss'
/storage/nnz/MOM_remapping_gpu_2020/build/../src/MOM_remapping.F90:1618:(.text+0xb611): relocation truncated to fit: R_X86_64_PC32 against `.bss'
/storage/nnz/MOM_remapping_gpu_2020/build/../src/MOM_remapping.F90:1647:(.text+0xb740): relocation truncated to fit: R_X86_64_32S against `.bss'
/storage/nnz/MOM_remapping_gpu_2020/build/../src/MOM_remapping.F90:1654:(.text+0xb8b3): relocation truncated to fit: R_X86_64_32S against `.bss'
/storage/nnz/MOM_remapping_gpu_2020/build/../src/MOM_remapping.F90:1666:(.text+0xbbb7): relocation truncated to fit: R_X86_64_32S against `.bss'
/storage/nnz/MOM_remapping_gpu_2020/build/../src/MOM_remapping.F90:1668:(.text+0xbc0c): relocation truncated to fit: R_X86_64_32S against `.bss'
/storage/nnz/MOM_remapping_gpu_2020/build/../src/MOM_remapping.F90:1679:(.text+0xc1bd): relocation truncated to fit: R_X86_64_PC32 against `.bss'
/storage/nnz/MOM_remapping_gpu_2020/build/../src/MOM_remapping.F90:1686:(.text+0xc6c4): relocation truncated to fit: R_X86_64_32S against `.bss'
/storage/nnz/MOM_remapping_gpu_2020/build/../src/MOM_remapping.F90:1692:(.text+0xc84c): additional relocation overflows omitted from the output
make: *** [prg_remap] Error 2
rm prg_unit_tests.o prg_remap.o
```

**Comment by nikizadehgfdl** [comment]
The poor man's compilation works and produces a.out, so the errors must be due to the fancy build process :)

```
%nvfortran prg_remap.F90 MOM_remapping.F90 MOM_error_handler.F90 MOM_string_functions.F90 regrid_edge_values.F90 PLM_functions.F90 regrid_solvers.F90 polynomial_functions.F90

% ./a.out 
 ==== MOM_remapping: remapping_unit_tests_gpu =================
time taken    9.593

```
BTW. curious that the test passes with pgi20 (nvfortran above) but fails with PGI18 (below):

```
% pgfortran prg_remap.F90 MOM_remapping.F90 MOM_error_handler.F90 MOM_string_functions.F90 regrid_edge_values.F90 PLM_functions.F90 regrid_solvers.F90 polynomial_functions.F90

% ./a.out 
 ==== MOM_remapping: remapping_unit_tests_gpu =================
time taken    9.670
 remapping_unit_tests_gpu: Failed remapping_core_w() with error:  
   0.3750000    

```

What does this mean?

**Comment by marshallward** [comment]
I build fine with `gfortran`, but cannot build with `nvfortran`.

If I build with `-r8` then I get the errors listed above.

If I take it out, then it seems to make progress but now I cannot find the `dumpgrid_`:
```
/home/u00u6fu5ih76xLCQKw357/MOM_remapping_gpu_2020/build/../src/prg_remap.F90:45: undefined reference to `dumpgrid_'
/home/u00u6fu5ih76xLCQKw357/MOM_remapping_gpu_2020/build/../src/prg_remap.F90:53: undefined reference to `dumpgrid_'
```
This seems to be a bug; `prg_remap.F90` is missing a `use` statement, but odd that `gfortran` did not raise an error.

There also appear to be some issues with parallel builds (`make -j`).

**Comment by nikizadehgfdl** [comment]
I added the following to prg_remap.F90 to bypass this
use MOM_remapping, only : dumpGrid

**Comment by marshallward** [comment]
OK, then the problem is the `-r8` flag.  This was added because there was a `1e-140` in the code, and Intel was raising a warning.  For whatever reason, `nvfortran` does not like this.

Will look into it when I get a chance.

**Comment by marshallward** [comment]
Using this flag seems to resolve the problem: `-mcmodel=medium`

Seems to be somehow related to 32 vs 64-bit addressing.  I don't know why enabling 8-byte reals would trigger this issue but this flag seems necessary.

Not sure how to integrate this into autoconf but perhaps a problem for another time.

**Comment by marshallward** [comment]
`-O2` also appears to "fix" this issue.

It is possible that optimization is pushing everything into the lower 2GB 32-bit address space, whereas non-optimized code requires 64-bit addressed data.

(Ed: autoconf's `FCFLAGS` only sets `-O2` as a default for GCC compilers, which is exacerbating the problem here for PGI.)

**Comment by marshallward** [comment]
Replacing all the fixed arrays with allocatables seems to prevent the compiler from putting any 64-bit data in the .bss, and we probably want to be testing against allocated arrays anyway, so I will send in this PR.

**Comment by adcroft** [comment]
I think this can be closed due to #20  but would rather someone who's tested do it

**Comment by marshallward** [comment]
I'm now building on the nvidia node without any fussing of flags.
